### ISSUE TITLE
Throws error when backward through sum() twice

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#13494]
## 🐛 Bug
`sum()` is special that we can backward() through it multiple times since it doesn't have saved variable, so nothing was really destroyed after the first backward.

This behavior is bad as we might silently backward through it multiple times without noticing, but the grad accumulates.

## To Reproduce
```
a = torch.randn(2, 2, 2, 2).requires_grad_(True)
a_sum = a.sum()
a_sum.backward()
a_sum.backward()
```

@zdevito told me briefly how to fix, I will send a PR for it soon.

**Comment by ailzhang** [comment]
Yea I understand this is valid from the memory effiency perspective, that doing so is totally fine.
But this behavior is not consistent with all other operators, I thought it was clearer if we throw an error for this special case and no confusion is made for users. What do you think @soumith @SsnL  @zdevito  ?

**Comment by SsnL** [comment]
The other operators error because they need the buffer that is freed, not because that users backprop more than once.

**Comment by soumith** [comment]
@ailzhang I actually think this should be the behavior that happens, so I dont think we should throw an error.
Maybe we can empathize with your perspective better if you explain in what case you had to deal with this.

**Comment by zdevito** [comment]
Ailing brought up a good point in person: typically when you accidentally have a gradient graph lying around that you have already used, you will get an error because you cannot backward twice.  This is a good error! You were most likely doing something wrong and if silently works, your grads will accumulate in unexpected cases. But in this particular case, the error never appears for basically implementation-defined reasons (at least, reasons that are more complicated than we expect our users to understand).

**Comment by zdevito** [comment]
I don't think we can expect our users to necessarily know when we will be and won't be saving buffers, so being inconsistent based on whether we actually have buffers can lead to surprising results.

**Comment by apaszke** [comment]
While I do agree the behavior is implementation dependent, this code is sipmly incorrect in the vast majority of cases, and I don’t think anyone will ever be hit by it. The “if you backward multiple times, grads get accumulated” invariant is still true, no matter if you ask us to retain buffers or not. It’s just that without those flags (which is not 100% correct) we can still satisfy your request and will proceed without an error. There’s nothin unexpected in it I’d say, and I don’t think spending time to do extra checks when they’re unnecessary is a good idea

**Comment by zdevito** [comment]
I'd expect that most people think that retain_graph=True is the only valid way to be allowed to call backward twice, and this behavior is an exception. If it were frequently occurring it would be really bad: someone would write a model relying on backwards to work multiple times, everything would work and then someone else would come along, replace one op for another, and get an error someone else in the code. 

However, since it requires you to only use things like sum and add to cause this corner case to happen, it is probably pretty rare for anyone to ever get into this situation in the first place. So it is probably not worth the added complexity of making the API consistent.

**Comment by ailzhang** [comment]
Thanks all for the discussion! :D 

To add a little context here, we discovered this behavior mismatch while comparing the sparse tensor backward to the dense one(cc: @weiyangfb ), and by using `%timeit` we were surprised to see that `sum` is allowed to backprop unlimited times but sparse sum throws an error, while all other operators don't have this mismatch.  We figured out the specialness of dense `sum` with help of @zdevito  and @goldsborough. We agreed that this might be a common sense that "Pytorch destroys something after the first backward, to keep the graph I have to do retain_graph", and this mismatched behavior could potentially cause some confusion to users. 

But I totally agree that this case is too rare to happen in real user code, and it doesn't worth the added complexity. Closing the issue.
